File: app/src/main/java/br/org/cesar/discordtime/stickysessions/data/remote/repository/NoteRemoteRepository.java
package br.org.cesar.discordtime.stickysessions.data.remote.repository;

import java.util.ArrayList;
import java.util.List;

import br.org.cesar.discordtime.stickysessions.BuildConfig;
import br.org.cesar.discordtime.stickysessions.data.remote.model.NoteRemote;
import br.org.cesar.discordtime.stickysessions.data.remote.service.NoteService;
import br.org.cesar.discordtime.stickysessions.data.repository.mapper.Mapper;
import br.org.cesar.discordtime.stickysessions.domain.model.Note;
import br.org.cesar.discordtime.stickysessions.domain.model.NoteFilter;
import br.org.cesar.discordtime.stickysessions.domain.repository.NoteRepository;
import io.reactivex.Single;

public class NoteRemoteRepository implements NoteRepository {

    private NoteService mNoteService;
    private Mapper<Note, NoteRemote> mMapper;

    public NoteRemoteRepository(NoteService service,  Mapper<Note, NoteRemote> mapper) {
        mNoteService = service;
        mMapper = mapper;
    }

    @Override
    public Single<Note> addNote(Note note) {
        NoteRemote noteRemoteToAdd = mMapper.mapFromDomain(note);

        return mNoteService.addNote(noteRemoteToAdd).map(
            noteRemoteReturned -> mMapper.mapToDomain(noteRemoteReturned)
        );
    }

    @Override
    public Single<List<Note>> listNotesForSession(NoteFilter noteFilter) {
        return mNoteService.listNotesForSession(noteFilter.idSession).map(
            noteRemotes -> {
                List<Note> mNotes = new ArrayList<>();

                for (NoteRemote noteRemote : noteRemotes) {
                    mNotes.add(mMapper.mapToDomain(noteRemote));
                }

                return mNotes;
            }
        );
    }

    @Override
    public Single<Boolean> removeNote(Note note) {
        return  mNoteService.removeNote(note.id).map(
            responseBody -> true
        );
    }
}

File: app/src/main/java/br/org/cesar/discordtime/stickysessions/data/remote/repository/NoteRemoteRepository.kt
package br.org.cesar.discordtime.stickysessions.data.remote.repository

import java.util.ArrayList

import br.org.cesar.discordtime.stickysessions.data.remote.model.NoteRemote
import br.org.cesar.discordtime.stickysessions.data.remote.service.NoteService
import br.org.cesar.discordtime.stickysessions.data.repository.mapper.Mapper
import br.org.cesar.discordtime.stickysessions.domain.model.Note
import br.org.cesar.discordtime.stickysessions.domain.model.NoteFilter
import br.org.cesar.discordtime.stickysessions.domain.repository.NoteRepository
import io.reactivex.Single

class NoteRemoteRepository(private val mNoteService: NoteService, private val mMapper: Mapper<Note, NoteRemote>) : NoteRepository {

    override fun addNote(note: Note): Single<Note> {
        val noteRemoteToAdd = mMapper.mapFromDomain(note)

        return mNoteService.addNote(noteRemoteToAdd).map { noteRemoteReturned -> mMapper.mapToDomain(noteRemoteReturned) }
    }

    override fun listNotesForSession(noteFilter: NoteFilter): Single<List<Note>> {
        return mNoteService.listNotesForSession(noteFilter.idSession).map { noteRemotes ->
            val mNotes = ArrayList<Note>()

            for (noteRemote in noteRemotes) {
                mNotes.add(mMapper.mapToDomain(noteRemote))
            }

            mNotes
        }
    }

    override fun removeNote(note: Note): Single<Boolean> {
        return mNoteService.removeNote(note.id).map { responseBody -> true }
    }
}
